plugins/embed-optimizer/class-embed-optimizer-tag-visitor.php

```diff
@@ -119,6 +158,80 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 			$this->added_lazy_script = true;
 		}
 
-		return true;
+		/*
+		 * At this point the tag is a figure.wp-block-embed, and we can return false because this does not need to be
+		 * measured and stored in URL Metrics. Only the child div.wp-block-embed__wrapper tag is measured and stored
+		 * so that this visitor can look up the height to set as a min-height on the figure.wp-block-embed. For more
+		 * information on what the return values mean for tag visitors, see <https://github.com/WordPress/performance/issues/1342>.
+		 */
+		return false;
+	}
+
+	/**
+	 * Gets the XPath for the embed wrapper.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context.
+	 * @return string XPath.
+	 */
+	private function get_embed_wrapper_xpath( OD_Tag_Visitor_Context $context ): string {
+		return $context->processor->get_xpath() . '/*[1][self::DIV]';
+	}
+
+	/**
+	 * Reduces layout shifts.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context.
+	 */
+	private function reduce_layout_shifts( OD_Tag_Visitor_Context $context ): void {
+		$processor           = $context->processor;
+		$embed_wrapper_xpath = $this->get_embed_wrapper_xpath( $context );
+
+		/**
+		 * Array of tuples of groups and their minimum heights keyed by the minimum viewport width.
+		 *
+		 * @var array<int, array{OD_URL_Metric_Group, int}> $group_minimum_heights
+		 */
+		$group_minimum_heights = array();
+
+		$denormalized_elements = $context->url_metric_group_collection->get_all_denormalized_elements()[ $embed_wrapper_xpath ] ?? array();
+		foreach ( $denormalized_elements as list( $group, $url_metric, $element ) ) {
+			if ( ! isset( $element['resizedBoundingClientRect'] ) ) {
+				continue;
+			}
+			$group_min_width = $group->get_minimum_viewport_width();
+			if ( ! isset( $group_minimum_heights[ $group_min_width ] ) ) {
+				$group_minimum_heights[ $group_min_width ] = array( $group, $element['resizedBoundingClientRect']['height'] );
+			} else {
+				$group_minimum_heights[ $group_min_width ][1] = min(
+					$group_minimum_heights[ $group_min_width ][1],
+					$element['resizedBoundingClientRect']['height']
+				);
+			}
```

Comment:
Can you change this so that each entry is an associative array with e.g. `group` and `minimum_height` keys? This makes the code more intuitive to follow.

Reply:
How about a390e15e919b267ee8fcc963351a2ac62dabd2b3?

Reply:
LGTM, thanks!